### PR TITLE
planner: refactor the join reorder codes

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29932

Problem Summary:
Refactor the codes in the join reorder algorithm for the later leading hints. 
Note: This PR doesn't contain any logical changes

### What is changed and how it works?
Refactor the codes in the join reorder algorithm for the later leading hints. 

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
All of the old test cases can pass.
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_join_reorder.go [hunk 1]
```diff
@@ -160,14 +210,8 @@ func (s *baseSingleGroupJoinOrderSolver) makeBushyJoin(cartesianJoinGroup []Logi
...
 				break
 			}
-			newJoin := s.newCartesianJoin(cartesianJoinGroup[i], cartesianJoinGroup[i+1])
-			for i := len(s.otherConds) - 1; i >= 0; i-- {
-				cols := expression.ExtractColumns(s.otherConds[i])
-				if newJoin.schema.ColumnsIndices(cols) != nil {
-					newJoin.OtherConditions = append(newJoin.OtherConditions, s.otherConds[i])
-					s.otherConds = append(s.otherConds[:i], s.otherConds[i+1:]...)
-				}
-			}
```
Comment (Reminiscent) [inline planner/core/rule_join_reorder.go:170]: This can be changed or not, what is your opinion? @qw4990 @winoros

**Comment by Reminiscent** [comment]
This refactoring doesn't add additional tests. The test goal is that we can pass the original tests. If you think we need more test cases, please comment

File: planner/core/rule_join_reorder.go (hunk 1 above)
Comment (qw4990) [reply planner/core/rule_join_reorder.go:170]: Actually, I'm not sure, but to make it easier to maintain, I think all join-creating operations should be handled by a standard and dedicated function, so I prefer to use `MakeJoin` here.

**Review by qw4990 — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**